### PR TITLE
Added Create Connector Step

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description

- [x] Added Create Connector Step
- [x] Resolved the static MLClient initialization and finally able to get the test working for register and deploy
- [x] Added failure tests for Register, Deploy steps.

### Issues Resolved
Fixes https://github.com/opensearch-project/opensearch-ai-flow-framework/issues/70

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                logger.info("Created connector successfully");
+                createConnectorFuture.complete(
+                    new WorkflowData(Map.ofEntries(Map.entry("connector-id", mlCreateConnectorResponse.getConnectorId())))
+                );
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error("Failed to create connector");
+                createConnectorFuture.completeExceptionally(new FlowFrameworkException(e.getMessage(), RestStatus.INTERNAL_SERVER_ERROR));
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java:76]: Let's use [ExceptionsHelper.status](https://github.com/opensearch-project/OpenSearch/blob/1e28738b8c966011bf1ae1f00431f0377761cb0a/libs/core/src/main/java/org/opensearch/ExceptionsHelper.java#L95) to determine the RestStatus code of the exception

File: src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        return NAME;
+    }
+
+    private static Map<String, String> getParameterMap(Map<String, String> params) {
+
+        Map<String, String> parameters = new HashMap<>();
+        for (String key : params.keySet()) {
+            String value = params.get(key);
+            try {
+                AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java:147]: Just trying to learn, why this special handling?

File: src/main/java/org/opensearch/flowframework/workflow/DeployModelStep.java [hunk 3]
```diff
@@ -57,8 +55,8 @@ public void onResponse(MLDeployModelResponse mlDeployModelResponse) {
 
             @Override
             public void onFailure(Exception e) {
-                logger.error("Model deployment failed");
-                deployModelFuture.completeExceptionally(e);
+                logger.error("Failed to deploy model");
+                deployModelFuture.completeExceptionally(new FlowFrameworkException(e.getMessage(), RestStatus.INTERNAL_SERVER_ERROR));
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/DeployModelStep.java:59]: Let's use `ExceptionsHelper.status` here too to set the RestStatus code based on the exception type

File: src/main/java/org/opensearch/flowframework/workflow/RegisterModelStep.java [hunk 4]
```diff
@@ -80,7 +77,7 @@ public void onResponse(MLRegisterModelResponse mlRegisterModelResponse) {
             @Override
             public void onFailure(Exception e) {
                 logger.error("Failed to register model");
-                registerModelFuture.completeExceptionally(new IOException("Failed to register model "));
+                registerModelFuture.completeExceptionally(new FlowFrameworkException(e.getMessage(), RestStatus.INTERNAL_SERVER_ERROR));
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/RegisterModelStep.java:80]: here as well

File: src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    @Override
+    public CompletableFuture<WorkflowData> execute(List<WorkflowData> data) throws IOException {
+        CompletableFuture<WorkflowData> createConnectorFuture = new CompletableFuture<>();
+
+        ActionListener<MLCreateConnectorResponse> actionListener = new ActionListener<>() {
+
+            @Override
+            public void onResponse(MLCreateConnectorResponse mlCreateConnectorResponse) {
+                logger.info("Created connector successfully");
+                createConnectorFuture.complete(
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java:68]: Can we add a TODO here to handle updating the global context entry created resources/output field. As well as for the Register and Deploy model step

**Review by joshpalis — COMMENTED** [review]
Initial pass

File: src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                    case ACTIONS_FIELD:
+                        actions = (List<ConnectorAction>) content.get(ACTIONS_FIELD);
+                        break;
+                }
+
+            }
+        }
+
+        if (Stream.of(name, description, version, protocol, parameters, credentials, actions).allMatch(x -> x != null)) {
+            MLCreateConnectorInput mlInput = MLCreateConnectorInput.builder()
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java:121]: We should complete exceptionally if any of these fields are null. The action listener wont be hit if we dont end up executing the create connector request

File: src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java (hunk 2 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java:148]: ml-commons does this special handling for the params map [here](https://github.com/opensearch-project/ml-commons/blob/main/common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java#L72. But now I'm thinking do we have to do it on our end as well. Probably not, so I'll go ahead and remove this and just type cast to `Map<String, String>`.

File: src/main/java/org/opensearch/flowframework/workflow/RegisterModelStep.java [hunk 7]
```diff
@@ -139,7 +136,7 @@ public void onFailure(Exception e) {
                 .connectorId(connectorId)
                 .build();
 
-            machineLearningNodeClient.register(mlInput, actionListener);
+            mlClient.register(mlInput, actionListener);
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/RegisterModelStep.java:139]: here as well

File: src/main/java/org/opensearch/flowframework/workflow/DeployModelStep.java [hunk 8]
```diff
@@ -70,7 +68,7 @@ public void onFailure(Exception e) {
                 break;
             }
         }
-        machineLearningNodeClient.deploy(modelId, actionListener);
+        mlClient.deploy(modelId, actionListener);
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/DeployModelStep.java:71]: additionally here, we should complete exceptionally if the `model_id` field is not present

**Review by joshpalis — COMMENTED** [review]
Second pass, besides these comments, looks good to me

**Review by dbwiddis — COMMENTED** [review]
LGTM, pending the unresolved comments.

**Review by amitgalitz — APPROVED** [review]

**Review by joshpalis — APPROVED** [review]
Thanks for the changes

**Pull request merged**